File: go/cote/LTCD/이진_탐색_트리_범위의_합.go
package LTCD

import (
	"fmt"
	"math/rand"
	"container/list"
)

// const (
// 	TEST_CASE = 10
// 	MAX_VALUE = 100
// 	MAX_DEPTH = 20
// )

type 이진_탐색_트리_범위의_합 struct {
	TEST_CASE int
	MAX_VALUE int
	MAX_DEPTH int
}

type Node2 struct {
	Value int
	Left *Node2
	Right *Node2
}

func (problem 이진_탐색_트리_범위의_합) initialize(depth int, left, right int, random *rand.Rand) *Node2 {
	if (left + 1 >= right){
		return nil
	}

	if depth >= problem.MAX_DEPTH {
		return nil
	}

	if random.Float32() / float32(depth) < 1.0 / float32(problem.MAX_DEPTH) {
		return nil
	}

	node := new(Node2)

	node.Value = random.Intn(right - left) + left
	node.Left = problem.initialize(depth + 1, left, node.Value - 1, random)
	node.Right = problem.initialize(depth + 1, node.Value, right, random)

	return node
}

func (problem 이진_탐색_트리_범위의_합) dfs(node * Node2, left, right int) int {
	if node == nil {
		return 0
	}

	if node.Value < left {
		return problem.dfs(node.Right, left, right)
	}

	if node.Value >= right {
		return problem.dfs(node.Left, left, right) 
	}

	return node.Value + problem.dfs(node.Left, left, right) + problem.dfs(node.Right, left, right) 
}

func (problem 이진_탐색_트리_범위의_합) solve1(node *Node2, left, right int) int {
	return problem.dfs(node, left, right)
}

func (problem 이진_탐색_트리_범위의_합) solve2(node *Node2, left, right int) int {

	queue := list.New()
	queue.PushBack(node)

	acc :=0
	for {
		if queue.Len() == 0 {
			break
		}

		cur := queue.Front().Value.(*Node)
		queue.Remove(queue.Front())

		if cur == nil {
			continue
		}

		if cur.Value > left {
			queue.PushBack(cur.Left)
		}

		if cur.Value < right {
			queue.PushBack(cur.Right)
		}

		if left <= cur.Value && cur.Value < right {
			acc += cur.Value
		}

	}

	return acc
}

func (problem 이진_탐색_트리_범위의_합) main() {
	random := rand.New(rand.NewSource(43))
	for it := 0; it < problem.TEST_CASE; it++ {
		node := problem.initialize(0, 0, problem.MAX_VALUE, random)
		right := random.Intn(problem.MAX_VALUE - 1) + 1
		left := random.Intn(right)
		fmt.Println(problem.solve1(node, left, right), problem.solve2(node, left, right))
	}
}

File: python/cote/LTCD/이진_탐색_트리_범위의_합.py
from __future__ import annotations
from typing import Generic, TypeVar
from collections import deque
import random

TEST_CASES = 10
MAX_VALUE = 100
MAX_DEPTH = 20

T = TypeVar("T")


class Node(Generic[T]):

    def __init__(self, value: T, left: Node[T] = None, right: Node[T] = None):
        self.value = value
        self.left = left
        self.right = right

    def __iter__(self):
        queue = deque([self])

        while queue:
            node = queue.popleft()

            if node:
                yield node.value
                queue.append(node.left)
                queue.append(node.right)
            else:
                yield None

    def __str__(self):
        return str([*self])


def solve(node: Node[int], left: int, right: int) -> int:

    def dfs(cur: Node[int]) -> int:
        if not cur:
            return 0
        if cur.value < left:
            return dfs(cur.right)
        if cur.value >= right:
            return dfs(cur.left)

        return cur.value + dfs(cur.left) + dfs(cur.right)

    return dfs(node)


def solve2(node: Node[int], left: int, right: int) -> int:
    queue = deque([node])
    acc = 0

    while queue:
        cur = queue.popleft()
        if not cur:
            continue
        if cur.value > left:
            queue.append(cur.left)
        if cur.value < right:
            queue.append(cur.right)
        if left <= cur.value < right:
            acc += cur.value

    return acc


def initialize(depth: int, left: int, right: int):
    if left + 1 >= right:
        return

    if depth >= MAX_DEPTH:
        return None

    if depth > 0 and random.random() / depth < 1 / MAX_DEPTH:
        return None

    mid = random.randrange(left + 1, right)

    left_node = initialize(depth + 1, left, mid)
    right_node = initialize(depth + 1, mid, right)

    return Node(mid, left_node, right_node)


if __name__ == "__main__":
    random.seed(43)

    for _ in range(TEST_CASES):
        node = initialize(0, 0, MAX_VALUE)
        right = random.randrange(1, MAX_VALUE)
        left = random.randrange(right)
        print(solve(node, left, right), solve2(node, left, right))
